tools/include/eosio/abigen.hpp

```diff
@@ -530,5 +626,143 @@ namespace eosio { namespace cdt {
          std::set<abi_table>                   ctables;
          std::map<std::string, std::string>    rcs;
          std::set<const clang::Type*>          evaluated;
+
+         bool is_kv_table(const clang::CXXRecordDecl* decl) {
+            for (const auto& base : decl->bases()) {
+               auto type = base.getType();
+               if (type.getAsString().find("eosio::kv_table<") != std::string::npos) {
+                  return true;
+               }
+            }
+
+            return false;
+         }
+
+         std::string get_value_from_decltype(const clang::DecltypeType* decl) {
+            if (const auto ref_type = dyn_cast<clang::LValueReferenceType>(decl->desugar())) {
```

Comment:
Do we only care about the case where the argument is an lvalue?  It is perfectly legal to call `decltype` on literals and rvalues and even functions.  

I am fine with the specificity of this function if:
  * it is named in a way that shows that it is not a general use function for getting the type from a decltype
  * we control the code on the CDT side that would hit this code and it wouldn't make much sense to ever use any of the above mentioned arguments (limit the chance of someone coming in later and causing ABI generation internal errors)

Reply:
We (kind of) control the code that hits this because its only called currently in the case where we hit an index that has been defined using the KV_INDEX macro, which will always be an lvalue ref. We could either:

1. Make this more generic to handle many different types of `decltype` calls
2. Rename this function to make it clear that its being used in the specific case mentioned above.

Reply:
I'm fine with just renaming the function for now.  If we need it to be more generic in the future then we can do that then.